test/widgets/compose_box_test.dart

```diff
@@ -1354,4 +1400,462 @@ void main() {
+        await tester.tap(find.widgetWithText(ZulipWebUiKitButton, 'Save'));
+        checkRequest(messageId, prevContent: 'foo', content: 'bar');
+        await tester.pump(Duration.zero);
+        checkNotInEditingMode(tester, narrow: narrow);
+      });
+    }
+    doTestInterruptComposingFromActionSheet(narrow: channelNarrow);
+    doTestInterruptComposingFromActionSheet(narrow: topicNarrow);
+    doTestInterruptComposingFromActionSheet(narrow: dmNarrow);
+
+    // Test the "Discard…?" confirmation dialog when there's text in the compose
+    // box for a new message, but you've said you want to edit a message.
+    void doTestInterruptComposingFromFailedEdit({required Narrow narrow}) {
+      testWidgets('interrupting new-message compose by tapping failed edit to restore: $narrow', (tester) async {
+        TypingNotifier.debugEnable = false;
+        addTearDown(TypingNotifier.debugReset);
+
+        final messageId = msgIdInNarrow(narrow);
+        await prepareEditMessage(tester, narrow: narrow);
+
+        await startInteractionFromActionSheet(tester,
+          messageId: messageId, originalRawContent: 'foo');
+        await tester.pump(Duration(seconds: 1)); // raw-content request
+        await enterContent(tester, 'bar');
+
+        connection.prepare(apiException: eg.apiBadRequest());
+        await tester.tap(find.widgetWithText(ZulipWebUiKitButton, 'Save'));
+        connection.takeRequests();
+        await tester.pump(Duration.zero);
+        checkNotInEditingMode(tester, narrow: narrow);
+        check(find.text('EDIT NOT SAVED')).findsOne();
+
+        await enterContent(tester, 'composing new message');
+
+        // Expect confirmation dialog; tap Cancel
+        await tester.tap(find.text('EDIT NOT SAVED'));
+        await tester.pump();
+        await expectAndHandleDiscardConfirmation(tester, shouldContinue: false);
+        checkNotInEditingMode(tester,
+          narrow: narrow, expectedContentText: 'composing new message');
+
+        // Twiddle the input to make sure it still works
+        await enterContent(tester, 'composing new message…');
+
+        // Try again, but this time tap Discard and expect to enter edit session
+        await tester.tap(find.text('EDIT NOT SAVED'));
+        await tester.pump();
+        await expectAndHandleDiscardConfirmation(tester, shouldContinue: true);
+        await tester.pump();
+        checkContentInputValue(tester, 'bar');
+        await enterContent(tester, 'baz');
+
+        // Save; check that the request is made and the compose box resets.
+        connection.prepare(json: UpdateMessageResult().toJson());
+        await tester.tap(find.widgetWithText(ZulipWebUiKitButton, 'Save'));
+        checkRequest(messageId, prevContent: 'foo', content: 'baz');
+        await tester.pump(Duration.zero);
+        checkNotInEditingMode(tester, narrow: narrow);
+      });
+    }
+    doTestInterruptComposingFromFailedEdit(narrow: channelNarrow);
```

Comment:
nit: for helpers that define a test case like `test` and `testWidgets` do, name them `testFoo` to match those:
```suggestion
    testInterruptComposingFromFailedEdit(narrow: channelNarrow);
```

We're mostly consistent about that practice already; for examples, see `git grep -P '\btest[^W]\w+\('`. (Or refine the pattern to possibly find more examples; it's excluding `testWidgets` but crudely so.)